File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not subRoot:
            return True
        if not root:
            return False
        
        if self.isIdentical(root, subRoot):
            return True
        
        return self.isIdentical(root.left, subRoot) or self.isIdentical(root.right, subRoot)
    
    def isIdentical(self, node1, node2):
        if not node1 and not node2:
            return True
        if not node1 or not node2 or node1.val != node2.val:
            return False
        
        return self.isIdentical(node1.left, node2.left) and self.isIdentical(node1.right, node2.right)

File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSubtree(root *TreeNode, subRoot *TreeNode) bool {
    if subRoot == nil {
        return true
    }

    if root == nil {
        return false
    }

    if isIdentical(root, subRoot) {
        return true
    }

    return isSubtree(root.Left, subRoot) || isSubtree(root.Right, subRoot)

}

func isIdentical(node1, node2 *TreeNode) bool {
    if node1 == nil && node2 == nil {
        return true
    }

    if node1 == nil || node2 == nil || node1.Val != node2.Val {
        return false
    }

    return isIdentical(node1.Left, node2.Left) && isIdentical(node1.Right, node2.Right)
}